runtime/runtime/src/tests/apply.rs

```diff
@@ -2793,3 +2837,169 @@ fn test_deploy_and_call_local_receipts() {
         ActionErrorKind::FunctionCallError(FunctionCallError::MethodResolveError(_))
     );
 }
+
+#[test]
+fn test_transaction_ordering_with_apply() {
+    let alice_key1 = InMemorySigner::test_signer(&alice_account());
+    let bob_signer = InMemorySigner::test_signer(&bob_account());
+    let alice_key2 = InMemorySigner::from_seed(alice_account(), KeyType::ED25519, "seed");
+
+    let tx1 = SignedTransaction::send_money(
+        1,
+        alice_account(),
+        alice_account(),
+        &alice_key2, // invalid access key, tx should be dropped
+        100,
+        CryptoHash::default(),
+    );
+    let tx2 = SignedTransaction::send_money(
+        1,
+        alice_account(),
+        alice_account(),
+        &alice_key1,
+        200,
+        CryptoHash::default(),
+    );
+    let tx3 = SignedTransaction::send_money(
+        2,
+        alice_account(),
+        bob_account(),
+        &alice_key1,
+        300,
+        CryptoHash::default(),
+    );
+    let tx4 = SignedTransaction::send_money(
+        1,
+        bob_account(),
+        bob_account(),
+        &bob_signer,
+        400,
+        CryptoHash::default(),
+    );
+    let tx5 = SignedTransaction::send_money(
+        2,
+        bob_account(),
+        alice_account(),
+        &bob_signer,
+        500,
+        CryptoHash::default(),
+    );
+    let tx6 = SignedTransaction::send_money(
+        3,
+        bob_account(),
+        bob_account(),
+        &bob_signer,
+        600,
+        CryptoHash::default(),
+    );
+
+    let txs = vec![tx4.clone(), tx1, tx2.clone(), tx5.clone(), tx3.clone(), tx6.clone()];
+
+    let (runtime, tries, root, apply_state, _signers, epoch_info_provider) = setup_runtime(
+        vec![alice_account(), bob_account()],
+        to_yocto(1_000_000),
+        to_yocto(500_000),
+        10u64.pow(15),
+    );
+
+    let flags = vec![true; txs.len()];
+    let s_valid_txs = SignedValidPeriodTransactions::new(&txs, flags.as_slice());
+    let apply_result = runtime
+        .apply(
+            tries.get_trie_for_shard(ShardUId::single_shard(), root),
+            &None,
+            &apply_state,
+            &[],
+            s_valid_txs,
+            &epoch_info_provider,
+            Default::default(),
+        )
+        .expect("apply should succeed");
+
+    let expected_order =
+        vec![tx4.get_hash(), tx2.get_hash(), tx5.get_hash(), tx3.get_hash(), tx6.get_hash()];
+
+    assert_eq!(
+        apply_result.outcomes.len(),
+        8,
+        "should have processed 5 transactions and 3 local receipts"
```

Comment:
I don't understand where the 3 local receipts came from.

Reply:
They are generated for transactions where `signer_id == receiver_id`, there are 3 such transactions in this test. I will add a note in a comments explaining that.